api/API_VERSIONING.md

```diff
@@ -0,0 +1,198 @@
+# API versioning guidelines
+
+The Envoy project (and in the future [UDPA](https://github.com/cncf/udpa)) takes API stability and
+versioning seriously. Providing stable APIs is a necessary step in ensuring API adoption and success
+of the ecosystem. Below we articulate the API versioning guidelines that aim to deliver this
+stability.
+
+# API semantic versioning
+
+The Envoy APIs consist of a family of packages, e.g. `envoy.admin.v2alpha`,
+`envoy.service.trace.v2`. Each package is independently versioned with a protobuf semantic
+versioning scheme based on https://cloud.google.com/apis/design/versioning.
+
+The major version for a package is captured in its name (and directory structure). E.g. version 2
+of the tracing API package is named `envoy.service.trace.v2` and its constituent protos are located
+in `api/envoy/service/trace/v2`. Every protobuf must live directly in a versioned package namespace,
+we do not allow subpackages such as `envoy.service.trace.v2.somethingelse`.
+
+Minor and patch versions will be implemented in the future, this effort is tracked in
+https://github.com/envoyproxy/envoy/issues/8416.
+
+In everyday discussion and GitHub labels, we refer to the `v2`, `v3`, `vN`, `...` APIs. This has a
+specific technical meaning. Any given message in the Envoy API, e.g. the `Bootstrap` at
+`envoy.config.bootstrap.v3.Boostrap`, will transitively reference a number of packages in the Envoy
+API. These may be at `vN`, `v(N-1)`, etc. The Envoy API is technically a DAG of versioned package
```

Comment:
I think whole paragraph might be usefully replaced with an example set of configurations.

Reply:
I think we'll have something more tangible to put here once we have the final cut of the `v3alpha` API and have updated the Envoy config examples, so I suggest punting until then.